url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C64Reader.java

```diff
@@ -0,0 +1,756 @@
+      }
+      for (int j = 0, m = dire.name.length(); j < m; j++) {
+        dir[p++] = (byte) dire.name.charAt(j);
+      }
+      // Fill up name...
+      for (int j = 0, m = 18 - dire.name.length(); j < m; j++) {
+        dir[p++] = ' ';
+      }
+      String type = dire.getTypeString();
+      for (int j = 0, m = type.length(); j < m; j++) {
+        dir[p++] = (byte) type.charAt(j);
+      }
+      dir[p++] = 0;
+    }
+
+    dir[p++] = 0;
+    dir[p++] = 0;
+
+    // Not inserted into memory yet...
+    try {
+      out.write(dir, 0, p);
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  // Returns filename of actual read file... or null if failed
+  public String readFile(String filename, int adr, OutputStream out) {
+    int sindex = 0;
+    DirEntry dire = null;
+
+    System.out.println("Loading: '" + filename + "' at " + adr);
+
+    if ((sindex = filename.indexOf('*')) >= 0) {
+      String match = filename.substring(0, sindex);
+      System.out.println("Matcher: " + match);
+      Enumeration names = dirEntries.keys();
+      while (names.hasMoreElements()) {
+        String name = (String) names.nextElement();
+        if (name.startsWith(match)) {
+          System.out.println("Found: " + name);
+          dire = (DirEntry) dirEntries.get(name);
+        }
+      }
+    } else if (filename.equals("$")) {
+      printDirListing(out);
+      return filename;
+    } else {
+      dire = (DirEntry) dirEntries.get(filename);
+      if (dire == null) {
+        for (int i = filename.length(), n = 16; i < n; i++) {
+          filename+=' ';
+        }
+        dire = (DirEntry) dirEntries.get(filename);
+      }
+    }
+
+    if (dire == null) return null;
+
+    if (type == TAPE) {
+      return readTapeFile(dire);
+    }
+
+    // Otherwise read the disk file!
+    return readDiskFile(dire, filename, adr, out);
+  }
+
+  public String readDiskFile(DirEntry dire, String str, int adr,
+      OutputStream out) {
+    byte[] sec = getSector(dire.trk, dire.sec);
+    // dumpSector(sec);
+    int address = (sec[2] & 0xff) + (sec[3] & 0xff) * 256;
+    System.out.println("*** Reading DISK file at " +
+        Integer.toString(address, 16));
+
+    if (out != null) {
+      // Send over the address if out exists - this is a part of the file!
+      try {
+        out.write(sec[2] & 0xff);
+        out.write(sec[3] & 0xff);
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    }
+
+    if (adr != -1) {
+      System.out.println("Address override: " + address + " -> " + adr);
+      address = adr;
+    }
+    int nextSector = sec[1] & 0xff;
+    int nextTrak = sec[0] & 0xff;
+
+    try {
+      for (int i = 0; i < 252; i++) {
+        if (out != null)
+          out.write(sec[i + 4] & 0xff);
+        else
+          memory[i + address] = sec[i + 4] & 0xff;
+      }
+    } catch (Exception e) {
+      System.out.println("Could not write to output stream");
+      e.printStackTrace();
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 379:389, 401:411.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.